Slim/Plugin/UPnP/MediaRenderer/AVTransport.pm

```diff
@@ -330,21 +376,56 @@ sub SetNextAVTransportURI {
 		return [ 718 => 'Invalid InstanceID' ];
 	}
 
- 		my $meta = $class->_DIDLLiteToHash( $args->{NextURIMetaData} );
+	my $playlist = Slim::Player::Playlist::playList($client);
+	my $pd = $client->pluginData();
+
+	if ( scalar @{$playlist} > 1 ){
+
+		# Next track queued on the control point has changed while
+		# currentURI is still playing. Remove it from the playlist
+		# ready for the updated NextURI.
+		$client->execute(["playlist", "delete", 1]);
+	}
+
+	if ( exists $args->{NextURI} && $args->{NextURI} eq '' ) {
+
+		# NextURI is blank, clear NextURI track from playlist
+		$pd->{avt_NextAVTransportURIMetaData_hash} = '';
+		main::DEBUGLOG && $log->is_debug && $log->debug("NextURI cleared");
+
+	} else {
+
+		my $meta = $class->_DIDLLiteToHash( $args->{NextURIMetaData} );
 
 		# XXX parse playlist?
 
 		if ( !$meta->{res} ) {
 			return [ 714 => 'Illegal MIME-type' ];
 		}
 
-			# Save hash version of metadata
-	$client->pluginData( avt_NextAVTransportURIMetaData_hash => $meta );
+		my $upnp_uri = $meta->{res}->{uri};
+		my $upnp_uricached = $pd->{avt_NextAVTransportURIMetaData_hash}->{res}->{uri};
+
+		# only update if NextURI has changed or it's not queued yet
+		if ( $upnp_uri ne $upnp_uricached || scalar @{$playlist} < 2 ){
+
+			# Convert URI to protocol handler
+			$upnp_uri =~ s/^http/upnp/;
+			Slim::Music::Info::setBitrate( $upnp_uri, $meta->{res}->{bitrate} );
+			Slim::Music::Info::setDuration( $upnp_uri, $meta->{res}->{secs} );		
+
+			$pd->{avt_NextAVTransportURIMetaData_hash} = $meta;
+
+			# use insert to make sure it is queued next.
+			$client->execute( [ 'playlist', 'insert', $upnp_uri, $meta->{title} ] );	
+			main::DEBUGLOG && $log->is_debug && $log->debug("NextURI set " . $upnp_uri . ":" . $meta->{title});
+		}
+	}
 
 	# Change state variables
 	$class->changeState( $client, {
-		NextAVTransportURI         => $args->{NextURI},
-		NextAVTransportURIMetaData => $args->{NextURIMetaData},
+		NextAVTransportURI			=> $args->{NextURI},
+		NextAVTransportURIMetaData	=> $args->{NextURIMetaData},
```

Comment:
Please do not apply whitespace only changes (or other "no-op" changes) - except if it's to fix invalid indentation. But aligning code using tabs doesn't work, as you can see here.